Gems/Atom/Tools/ShaderManagementConsole/Code/Source/Document/ShaderManagementConsoleDocument.cpp

```diff
@@ -60,6 +62,77 @@ namespace ShaderManagementConsole
+    {
+        auto shaderAssetResult =
+            AZ::RPI::AssetUtils::LoadAsset<AZ::RPI::ShaderAsset>(m_absolutePath,
+                                                                 m_shaderVariantListSourceData.m_shaderFilePath);
+        if (shaderAssetResult)
+
+        {
+            AZ::Data::Asset<AZ::RPI::ShaderAsset> shaderAsset = shaderAssetResult.GetValue();
+            AZStd::vector<AZ::RPI::ShaderOptionDescriptor> options = shaderAsset->GetShaderOptionGroupLayout()->GetShaderOptions();
+            AZ::RPI::ShaderVariantListSourceData::VariantInfo variantInfo;
+            for (auto& optionDescriptor : options)
+            {
+                variantInfo.m_options[optionDescriptor.GetName()] = ""; // all unset
+            }
+            variantInfo.m_stableId = m_shaderVariantListSourceData.m_shaderVariants.empty()
+                ? 0
+                : m_shaderVariantListSourceData.m_shaderVariants.back().m_stableId + 1;
+            m_shaderVariantListSourceData.m_shaderVariants.push_back(variantInfo);
+            return variantInfo.m_stableId;
+        }
+        return {};
+    }
+
+    void ShaderManagementConsoleDocument::AppendSparseVariantSet(
+        AZStd::vector<AZ::Name> optionHeaders,
+        AZStd::vector<AZ::Name> matrixOfValues)
+    {
+        // Make a lookup table to "reverse" the vector given to us in argument
+        AZStd::unordered_map<AZ::Name, int> nameToHeaderIndex;
+        for (int i = 0; i < optionHeaders.size(); ++i)
+        {
+            nameToHeaderIndex[optionHeaders[i]] = i;
+        }
+        // Prepare a whole new source data
+        AZ::RPI::ShaderVariantListSourceData newSourceData{ m_shaderVariantListSourceData };
+        AZ::u32 stableId = newSourceData.m_shaderVariants.empty() ? 1 : newSourceData.m_shaderVariants.back().m_stableId + 1;
+        if (matrixOfValues.size() % optionHeaders.size() != 0)
+        {
+            AZ_Error("ShaderManagementConsoleDocument", false,
+                "AppendSpareseVariantSet: matrixOfValues size must be multiple of header count");
+            return;
+        }
+        // add "line by line"
+        int numLines = matrixOfValues.size() / optionHeaders.size();
+        for (int line = 0; line < numLines; ++line)
+        {
+            AZ::RPI::ShaderOptionValuesSourceData mapOfOptionNameToValues;
+            size_t count = GetShaderOptionDescriptorCount();
+            // we need to fill-in the gaps by adding empty values for unset options, so we loop on all descriptors
+            for (int column = 0; column < count; ++column)
+            {
+                auto& descriptor = GetShaderOptionDescriptor(column);
+                auto& optionName = descriptor.GetName();
+                auto indexIt = nameToHeaderIndex.find(optionName);
+                if (indexIt == nameToHeaderIndex.end())
+                {
+                    mapOfOptionNameToValues[optionName] = ""; // unset
+                }
+                else
```

Comment:
I was wondering about why we need to do this as well, but it appeared to me we don't have the descriptors actually. So that's why the RebuildTable system can't know about options that have no keys in the map.
We don't really conserve them after document opening. This is the reason why `AddOneVariantRow` needs to re-access the original shader asset using LoadAsset which I suspect isn't efficient.
Personally, I would prefer the behavior to be like you described. We would probably just need to save a `AZStd::vector<AZ::RPI::ShaderOptionDescriptor> m_descriptors;` somewhere in the document

Reply:
The descriptors are accessible from the shader asset.
The shader management console document loads the shader asset as part of opening a shader or shader variant list file.
This is enforced when calling ShaderManagementConsoleDocument::SetShaderVariantListSourceData, which does some additional validation and uses the descriptor to retrieve default values and prune empty options.
In an earlier comment, I think I recommended rewriting some of these functions in terms of ShaderManagementConsoleDocument::SetShaderVariantListSourceData.
For example, the add one row function could simply add the row directly to the shader variant list source data with the unique option ID, then call SetShaderVariantListSourceData. This might seem like overkill but should cover most of the edge cases and other validation. 
All of the table view logic was previously making changes to a local copy of the shader variant list source data then calling this function to update the document.
Either way, it can be addressed after the fact but I remember issues and comments from past PRs when I created the table view and originally moved everything from Python to C++.

Reply:
Your remark is correct as far as I know. What I'm concerned with is the opening of the shader asset, it's probably not that heavy but the console itself is already super slow, for example the python script that generates enumerations can prepare the enumerated table in memory for tens of thousands of variants (using the embedded pycharm interpreter) in a split second. But adding the lines in Qt takes hours. I left it run for 4 hours and it got to ~9000 table rows only (in profile mode).
We have serious performance problems so I'd rather not make it worse by opening the shader asset at each little delta on the data model. Maybe if we need it, we can keep a cached version of the descriptor vector.

Reply:
Yes but it opens it only when the list is empty, I'm thinking it was intended as a detector for "the first time".
We should revisit it because it's not a very robust detector, we can cache what's necessary, or keep the whole asset itself as a memory object copy, I don't think it's heavy.